### ISSUE TITLE
Access full class name in `CoproductHint` 

### ISSUE THREAD

**Issue by ruippeixotog** [fommil/spray-json-shapeless#26]
I am dealing with a message protocol in which message types are identified in JSON by two keys: one that identifies a subprotocol (`protocol`) and another that identifies a command in that subprotocol (`command`). The same `command` value may appear in multiple subprotocols with different meanings.

My first approach was to create a class structure like this:
 
```scala
import spray.json._
import fommil.sjs.FamilyFormats._

object Test extends App {
  sealed trait Message

  object ProtocolA {
    case class Start(id: String) extends Message
  }

  object ProtocolB {
    case class Start(id: String) extends Message
  }

  // prints: { "type": "Start", "id": "abc" }
  // wanted: { "protocol": "ProtocolA", "type": "Start", "id": "abc" }
  println((ProtocolA.Start("abc"): Message).toJson)

  // prints: { "type": "Start", "id": "abc" }
  // wanted: { "protocol": "ProtocolB", "type": "Start", "id": "abc" }
  println((ProtocolB.Start("abc"): Message).toJson)
}
```

It's easy to create my custom `CoproductHint` to use the `command` JSON key instead of `type` and optionally do transformations on the class names. However, in this case it would be useful for me to access the fully qualified name of the class to be serialized, so I could extract the enclosing object to get the subprotocol.

Another alternative would be to encode both the subprotocol and the command in the class name (e.g. separated with an underscore) and override all methods in the custom `CoproductHint`, but this would result in weird class names for my models, I thing I would rather avoid.

Do you have any idea on how I could do this in an easier manner? Am I approaching the problem in a wrong way?

**Comment by fommil** [comment]
I'd model each message as a sealed trait that can be either protocol or command, each with their own subtype. You'll certainly need to customise the coproduct hint to do that. Does this make sense?

**Comment by ruippeixotog** [comment]
I didn't get the idea, I'm sorry...  I can't see how having a separate type hierarchy for protocols and commands would help me create a suitable `CoproductHint`. Can you elaborate a bit more?

**Comment by fommil** [comment]
Coproduct is for one or the other. So is perfect for what you want.

**Comment by ruippeixotog** [comment]
I know what is a coproduct, and it's clear from my comment that it was not what I was struggling with... I was looking for a much more practical hint on how to access the information I want inside `CoproductHint`(s) without recurring to what seems to me like hacky encodings of the class hierarchy or the JSON serialization.

For anyone having this problem, I ended up writing it [this way](https://gist.github.com/ruippeixotog/7f28a3d9df933a1cd80d15d466c0fbe1). I don't like the outer case class hierarchy, as I had to create an artificial `payload` field to hold the variable part of the messages which I have to wrap and unwrap in the `CoproductHint`. This also has the downside that the default `JsonFormat[Message]` is very different from the formats of its subclasses. @fommil, if you were thinking of a different solution or know some way to improve this which you want to share, I'd be happy to improve this.
